src/components/issue-event-list-item.component.js

```diff
@@ -0,0 +1,617 @@
+              {action}
+            </Text>
+            <LabelButton label={label} />
+          </View>
+          <Date date={createdAt} />
+        </View>
+      </View>
+    );
+  }
+}
+
+class LabelGroup extends Component {
+  props: {
+    group: Object,
+    onPressUser: Function,
+  };
+
+  render() {
+    const {
+      actor,
+      labeled,
+      unlabeled,
+      created_at: createdAt,
+    } = this.props.group;
+
+    const toInlineLabel = (type, { label }, index) =>
+      <InlineLabel key={type + index} label={label} />;
+
+    /* eslint-disable react/jsx-no-bind */
+    const labels = labeled.map(toInlineLabel.bind(null, 'added'));
+    const unlabels = unlabeled.map(toInlineLabel.bind(null, 'removed'));
+
+    let textChildren = [
+      <ActorLink actor={actor} onPress={this.props.onPressUser} />,
+    ];
+
+    if (labels.length) {
+      textChildren = [...textChildren, <Text> added </Text>, ...labels];
+    }
+
+    if (labels.length && unlabels.length) {
+      textChildren.push(<Text> and</Text>);
+    }
+
+    if (unlabels.length) {
+      textChildren = [...textChildren, <Text> removed </Text>, ...unlabels];
+    }
+
+    return (
+      <View style={[styles.container, styles.header]}>
+        <EventIcon name="tag" />
+        <View style={styles.contentContainer}>
+          <View style={styles.eventTextContainer}>
+            {textChildren}
+          </View>
+          <Date date={createdAt} />
+        </View>
+      </View>
+    );
+  }
+}
+
+class Assigned extends Component {
+  props: {
+    event: Object,
+    unassigned: boolean,
+    onPressUser: Function,
+  };
+
+  render() {
+    const { assigner, assignee, created_at: createdAt } = this.props.event;
+    const action = this.props.unassigned ? 'unassigned' : 'assigned';
+
+    return (
+      <Event
+        icon={<EventIcon name="person" />}
+        text={
+          <Text style={{ padding: 3 }}>
+            <ActorLink actor={assigner} onPress={this.props.onPressUser} />{' '}
+            {action}{' '}
+            <ActorLink actor={assignee} onPress={this.props.onPressUser} />
+          </Text>
+        }
+        createdAt={createdAt}
+      />
+    );
+  }
+}
+
+class Milestoned extends Component {
+  props: {
+    event: Object,
+    demilestoned: boolean,
+    onPressUser: Function,
+  };
+
+  render() {
+    const { actor, milestone, created_at: createdAt } = this.props.event;
+    const action = this.props.demilestoned
+      ? 'removed this from'
+      : 'added this to';
+
+    return (
+      <Event
+        icon={<EventIcon name="milestone" />}
+        text={
+          <Text style={{ padding: 3 }}>
+            <ActorLink actor={actor} onPress={this.props.onPressUser} />{' '}
+            {action} the <Bold>{milestone.title}</Bold> milestone
+          </Text>
+        }
+        createdAt={createdAt}
+      />
+    );
+  }
+}
+
+class Locked extends Component {
```

Comment:
I like how we have a base `Event` component that's being used for other scenarios, but still think we can probably simplify things here a bit. Instead of having each component here take in the event and decide which icon to use, etc.., we can probably extend this with props. For example: 

```
class EventInstance extends Component {
  props: {
    icon: string,
    iconColor: color,
    createdAt: string,
    textChildren: React.Element<*>,
  };

  render() {
    const {
      name,
      iconColor = '#586069',
      iconBackgroundColor = '#E6EBF1',
      createdAt,
      textChildren
    } = this.props;
    const action = this.props.unassigned ? 'unassigned' : 'assigned';

    return (
      <Event
        icon={
          <Icon
            iconStyle={{
              marginLeft: marginLeftForIconName(name),
              marginTop: 1,
              color: iconColor,
            }}
            containerStyle={[styles.iconContainer, { iconBackgroundColor }]}
            name={icon}
            type="octicon"
            size={16}
          />
        }
        text={
          <Text style={{ padding: 3 }}>
            {textChildren}
          </Text>
        }
        createdAt={createdAt}
      />
    );
  }
}
```

If you think it makes sense to have a `Event` component specified on it's own, then we can probably use something like this `EventInstance` everywhere instead of having `Locked`, `Headref`, etc.. which are essentially the same components with slight variations. Please let me know what you think.

Reply:
Notice how I'm just referencing `Icon` directly. Don't know if it's worth creating an `EventIcon` if we use a component like the above everywhere in my opinion as it'll only be used in one place (this component)

Reply:
i think i see what you're saying. i'm happy to rearrange things, but i want to make my case first 😄. i'm not trying to argue or to even say that things are ideal how they're currently implemented, but i want to understand more of the _why_ you'd prefer things to be structured differently.

the more `react` i've written, the more i've tended to break things apart into a bunch of small components. these components can be composed together to build larger structures.

you seem to be somewhat against tiny components like `Bold`. in several different places surrounding inline text, i could write `<Text style={styles.boldText}>` or i could pull that out into `<Bold>`, allowing me to write nice things like

```js
changed the title from <Bold>{event.rename.from.trim()}</Bold>{' '}
to <Bold>{event.rename.to.trim()}</Bold>
```

isn't this encapsulation the point of react?

similarly, i _could_ get rid of `<EventIcon>`, but even if it's only in like 2 or 3 spots, i'd always be setting the same `name`, `size`, and `containerStyle` on an `Icon`. why not pull it out?

as a project grows, i tend to build up a library of these small components. even if the small components are relatively context specific, refactoring tends to feel easy, cause i can often just build a new component using a few existing building blocks and be done with it.

y'all have invested way more into this project than i have, so i'm happy to structure things as you'd like. it just feels as tho i'd end up with more duplication and larger, more complex `render` functions, neither of which i'm a big fan of. however, i could just be misunderstanding you or missing something altogether.

thanks again!!!

Reply:
Oh no please don't ever hesitate to let us know what you're thinking mate :)

Yeah everyone prefers to design their component renders in different fashions. I just happen to lean towards creating sub-components _if_ I feel like they could live as separate entity in the components directory and used throughout the app. But you're right, there's definitely benefit to break things down within the same component file and not have large render methods spread out across the app.

Definitely see where you're coming from trying to arrange things the way I suggested in causing large, duplicated logic. Perfectly fine with the way you've set things up if you feel like this is a nicer way to structure things in the file :+1:

Reply:
i refactored some things in the direction you suggested, and i really like how things ended up! 

`<Event>` became a bit more powerful, and i was able to eliminate most of the smaller components. `<LabelGroup>` has to make quite a few decisions, so it held on. `<Bold>` and `<ActorLink>` still feel useful, so i kept them around too.

things feel more simple! let me know what you think 💯